tasks/message/watch_eth.go

```diff
@@ -0,0 +1,191 @@
+package message
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/lib/chainsched"
+
+	types2 "github.com/filecoin-project/lotus/chain/types"
+)
+
+type MessageWatcherEth struct {
+	db  *harmonydb.DB
+	ht  *harmonytask.TaskEngine
+	api *ethclient.Client
+
+	stopping, stopped chan struct{}
+
+	updateCh        chan struct{}
+	bestBlockNumber atomic.Pointer[big.Int]
+}
+
+func NewMessageWatcherEth(db *harmonydb.DB, ht *harmonytask.TaskEngine, pcs *chainsched.CurioChainSched, api *ethclient.Client) (*MessageWatcherEth, error) {
+	mw := &MessageWatcherEth{
+		db:       db,
+		ht:       ht,
+		api:      api,
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+		updateCh: make(chan struct{}, 1),
+	}
+	go mw.run()
+	if err := pcs.AddHandler(mw.processHeadChange); err != nil {
+		return nil, err
+	}
+	return mw, nil
+}
+
+func (mw *MessageWatcherEth) run() {
+	defer close(mw.stopped)
+
+	for {
+		select {
+		case <-mw.stopping:
+			// TODO: cleanup assignments
+			return
+		case <-mw.updateCh:
+			mw.update()
+		}
+	}
+}
+
+func (mw *MessageWatcherEth) update() {
+	ctx := context.Background()
+
+	bestBlockNumber := mw.bestBlockNumber.Load()
+
+	confirmedBlockNumber := new(big.Int).Sub(bestBlockNumber, big.NewInt(MinConfidence))
+	if confirmedBlockNumber.Sign() < 0 {
+		// Not enough blocks yet
+		return
+	}
+
+	machineID := mw.ht.ResourcesAvailable().MachineID
+
+	// Assign pending transactions with null owner to ourselves
+	{
+		n, err := mw.db.Exec(ctx, `UPDATE message_waits_eth SET waiter_machine_id = $1 WHERE waiter_machine_id IS NULL AND tx_status = 'pending'`, machineID)
```

Comment:
Could use an index on this table like https://github.com/filecoin-project/curio/pull/392/files